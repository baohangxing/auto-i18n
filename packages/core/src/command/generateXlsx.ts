import fsExtra from 'fs-extra'
import { getJsonPath, isUnTransed } from '../utils/config'
import { KEY_SYMBOL_IN_XLSX, XLSX_NAME } from '../utils/constants'
import { writeXlsxFile } from '../utils/excel'
import { createFileName, getKeys, getValueByKey } from '../utils/help'

const { readJsonSync } = fsExtra

const generateXlsx = async () => {
  const { baseLangJson, otherLangJsons } = getJsonPath()

  const baseLangJsonObj = readJsonSync(baseLangJson.path)

  const baseJsonKeySet = new Set(getKeys(baseLangJsonObj))

  const firstRows: string[] = [KEY_SYMBOL_IN_XLSX]
  const sheetsDatas: string[][] = []

  firstRows.push(baseLangJson.name)

  const langJsonObjMap = {} as any

  for (const lang of otherLangJsons) {
    firstRows.push(lang.name)
    langJsonObjMap[lang.name] = readJsonSync(lang.path)
  }

  for (const key of baseJsonKeySet) {
    const newRow: string[] = []
    newRow.push(key)
    newRow.push(getValueByKey(baseLangJsonObj, key))

    for (const lang of otherLangJsons) {
      const langVal = getValueByKey(langJsonObjMap[lang.name], key)
      newRow.push(isUnTransed(langVal, lang.name) ? '' : langVal)
    }

    sheetsDatas.push(newRow)
  }

  writeXlsxFile([firstRows], ['gm'], [sheetsDatas], `${createFileName(XLSX_NAME)}.xlsx`)
}

export { generateXlsx }
