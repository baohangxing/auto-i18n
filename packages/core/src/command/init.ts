import path from 'path'
import { cosmiconfigSync } from 'cosmiconfig'
import fs from 'fs-extra'
import serialize from 'serialize-javascript'
import { CLI_CONFIG_NAME } from '../utils/constants'
import log from '../utils/log'
import defaultAutoConfig from '../utils/defaultAutoConfig'
import { format } from '../utils/help'

const init = async () => {
  const explorer = cosmiconfigSync(CLI_CONFIG_NAME)
  const result = explorer.search()

  if (result?.config) {
    log.info('the project had inited, so skipped')
  }
  else {
    const configPath = path.resolve(process.cwd(), `${CLI_CONFIG_NAME}.config.cjs`)
    const code = `
    module.exports = ${serialize(defaultAutoConfig, {
      unsafe: true,
    })}
  `
    fs.outputFileSync(configPath, code)
    await format(configPath)
    log.success('inited')
  }
}

export { init }
