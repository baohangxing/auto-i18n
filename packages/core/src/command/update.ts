import path from 'path'
import fsExtra from 'fs-extra'
import {
  getKeys, getValueByKey, setValueByKey,
  sortObjectKey,
  writeToJsonFile,
} from '../utils/help'
import { getAutoConfig, getJsonPath, isUnTransed, unTransLocale } from '../config/config'
import log from '../utils/log'

import { readXlsxFile } from '../utils/excel'
import { KEY_SYMBOL_IN_XLSX } from '../config/constants'

const { readJsonSync } = fsExtra

const updateBaseLocale = async (replaceObj?: any) => {
  const { baseLangJson } = getJsonPath()
  let baseLangJsonObj

  if (replaceObj) {
    baseLangJsonObj = {}
    const addKeySet = new Set(getKeys(replaceObj))

    for (const k of addKeySet) {
      const v = getValueByKey(replaceObj, k)
      setValueByKey(baseLangJsonObj, k, v)
    }
  }
  else {
    baseLangJsonObj = readJsonSync(baseLangJson.path)
  }
  await writeToJsonFile(path.parse(baseLangJson.path).dir, baseLangJson.name, baseLangJsonObj)
}

const updateLocales = async () => {
  const autoConfig = getAutoConfig()
  const { baseLangJson, otherLangJsons } = getJsonPath()
  const baseLangJsonObj = readJsonSync(baseLangJson.path)

  await updateBaseLocale()

  for (const langJson of otherLangJsons) {
    const langJsonObj = readJsonSync(langJson.path)
    const baseJsonKeySet = new Set(getKeys(baseLangJsonObj))
    const langJsonKeySet = new Set(getKeys(langJsonObj))
    const newObj = sortObjectKey(langJsonObj)
    for (const x of baseJsonKeySet) {
      const baseValue = getValueByKey(baseLangJsonObj, x)
      const langValue = getValueByKey(langJsonObj, x)

      if (langValue === undefined
        || langValue === null
        || langValue === ''
        || isUnTransed(langValue, langJson.name)) {
        let transedWord = ''
        if (autoConfig.transLacaleWord) {
          transedWord = await autoConfig.transLacaleWord(
            baseValue,
            baseLangJson.name,
            langJson.name,
          )
        }
        else {
          transedWord = unTransLocale(baseValue, langJson.name)
        }
        setValueByKey(newObj, x, transedWord)
      }
      langJsonKeySet.delete(x)
    }
    if (langJsonKeySet.size > 0) {
      for (const x of langJsonKeySet) {
        log.info(
          `${x}: ${getValueByKey(langJsonObj, x)} in ${langJson.name
          }.json is deleted, Pleace check it is used in project!`,
        )
      }
    }

    await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newObj)
  }

  log.success('updateLocales success')
}

const updateLocalesFromXlsx = async (filePath: string) => {
  const res = readXlsxFile(path.join(path.resolve(), filePath))

  const { baseLangJson, otherLangJsons } = getJsonPath()
  if (res) {
    const baseLangJsonObj = readJsonSync(baseLangJson.path)

    for (const langJson of otherLangJsons) {
      const langJsonObj = readJsonSync(langJson.path)
      const typeJsonKeySet = new Set(getKeys(baseLangJsonObj))
      const newJsonObj = sortObjectKey(langJsonObj)

      for (const { data } of res) {
        const keyIndex = data[0].indexOf(KEY_SYMBOL_IN_XLSX)
        const langIndex = data[0].indexOf(langJson.name)

        if (keyIndex !== -1 && langIndex !== -1) {
          for (const aLine of data.slice(1)) {
            if (typeJsonKeySet.has(aLine[keyIndex])) {
              if (aLine[langIndex]) {
                setValueByKey(
                  newJsonObj,
                  aLine[keyIndex],
                  aLine[langIndex].toString(),
                )
              }
            }
          }
        }
      }
      await writeToJsonFile(path.parse(langJson.path).dir, langJson.name, newJsonObj)
      log.success(`Write to Json file success in ${langJson.path}`)
    }
  }
}

export { updateLocalesFromXlsx, updateLocales, updateBaseLocale }
