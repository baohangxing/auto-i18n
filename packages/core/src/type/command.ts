export interface TransCommandOption {
  filePath: string
  generateNewFile: boolean
  newFileName: string
}
