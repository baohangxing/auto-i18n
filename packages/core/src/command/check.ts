import path from 'path'
import fs from 'fs'
import fsExtra from 'fs-extra'
import { checkTranslated } from '../check'
import { getAutoConfig, getJsonPath, getOutputFileDir, isUnTransed } from '../config/config'
import type { FileExtension } from '../types'
import { fgSync } from '../utils/glob'
import { writeXlsxFile } from '../utils/excel'
import { createFileName, getKeys, getValueByKey } from '../utils/help'
import log from '../utils/log'

const checkAllTranslated = () => {
  const autoConfig = getAutoConfig()

  const { baseLangJson, otherLangJsons } = getJsonPath()

  let paths: string[] = []

  paths = fgSync(autoConfig.includes)

  const firstRows: string[][] = [
    ['paths', 'untrans key'],
    ['used path', 'keys num', 'keys'],
    ['used key', 'used paths'],
  ]
  const sheetsDatas: string[][][] = [[], [], []]

  const baseLangJsonObj = fsExtra.readJsonSync(baseLangJson.path)

  const allKeysSet = new Set(getKeys(baseLangJsonObj))

  const keysUsedSet = new Set<string>()

  const keysUsedMap = new Map<string, string[]>()

  const langJsonObjMap = {} as any

  firstRows[0].push(baseLangJson.name)

  for (const lang of otherLangJsons) {
    firstRows[0].push(lang.name)
    langJsonObjMap[lang.name] = fsExtra.readJsonSync(lang.path)
  }

  for (const filePath of paths) {
    const autoConfig = getAutoConfig()
    const ext = path.parse(filePath).ext.slice(1) as FileExtension
    const source = fs.readFileSync(filePath, 'utf8')

    const keys: string[] = []
    try {
      keys.push(...checkTranslated(source, ext, autoConfig.i18nCallRules))
    }
    catch (e) {
      log.error(`checkTranslated ${filePath} Error`, e)
      continue
    }

    for (const x of keys) {
      keysUsedSet.add(x)
      keysUsedMap.set(x, [...(keysUsedMap.get(x) ?? []), filePath])
    }
    sheetsDatas[1].push([filePath, keys.length.toString(), keys.join('\n\r')])
  }

  for (const key of keysUsedSet) {
    if (!allKeysSet.has(key))
      log.error(`json do not have key ${key}, pleace check ${baseLangJson.path}`)
    let unTransed = false
    for (const lang of otherLangJsons) {
      const langVal = getValueByKey(langJsonObjMap[lang.name], key)
      if (isUnTransed(langVal, lang.name))
        unTransed = true
    }

    if (unTransed) {
      const item: string[] = [keysUsedMap.get(key)?.join('\n\r') ?? '', key]

      item.push(getValueByKey(baseLangJsonObj, key))

      for (const lang of otherLangJsons) {
        const langVal = getValueByKey(langJsonObjMap[lang.name], key)
        item.push(langVal)
      }

      sheetsDatas[0].push(item)
    }

    sheetsDatas[2].push([key, keysUsedMap.get(key)?.join('\n\r') ?? ''])
  }

  writeXlsxFile(firstRows, ['unTrans key', 'all key detail', 'all key used'], sheetsDatas,
    getOutputFileDir(`${createFileName(autoConfig.outputXlsxNameBy.check)}.xlsx`))

  log.success('check successed')
}

export { checkAllTranslated }
