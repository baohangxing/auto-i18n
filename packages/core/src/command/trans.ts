import path from 'path'
import fs from 'fs'
import type { FileExtension, TransCommandOption } from '../types/config'
import log from '../utils/log'
import { transform } from '../transform'
import { getAutoConfig } from '../config/config'

const transformSingle = (filePath: string) => {
  const autoConfig = getAutoConfig()
  const ext = path.parse(filePath).ext.slice(1) as FileExtension
  const source = fs.readFileSync(filePath, 'utf8')
  const { code } = transform(source, ext, autoConfig.i18nCallRules)
  return code
}

const trans = (option: TransCommandOption) => {
  console.log(option)
  if (option.transPath) {
    const filePath = path.resolve(process.cwd(), option.transPath)
    const code = transformSingle(filePath)
    log.info(code)
  }
  else {
    // TODO
    log.info('TODO')
  }
}

export { trans }
