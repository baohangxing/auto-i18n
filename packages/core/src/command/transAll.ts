import fs from 'fs'
import path from 'path'
import { generateSFC } from 'vue-sfc-gen'
import { parse } from '@vue/compiler-sfc'
import consola from 'consola'
import type { TransCommandOption } from '../types'

const trans = (option: TransCommandOption) => {
  const filePath = path.resolve(option.filePath)

  if (fs.existsSync(filePath)) {
    const fileContent = fs.readFileSync(path.resolve(option.filePath)).toString()
    const res = generateSFC(parse(fileContent).descriptor)

    if (option.generateNewFile) {
      const { ext, dir, name } = path.parse(filePath)
      fs.writeFileSync(path.join(dir, `${option.newFileName ? option.newFileName : `${name}_transed`}${ext}`), res)
    }
    else {
      consola.success(`${fileContent} transed:`)
      consola.log(res)
    }
  }
  else { consola.error(`no this file: ${filePath}`) }
}

export { trans }
