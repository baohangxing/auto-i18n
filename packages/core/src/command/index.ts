import { Command, Option } from 'commander'
import { CLI_CONFIG_NAME } from '../config/constants'
import type { RevertCommandOption, TransCommandOption } from '../types/config'
import { checkAllTranslated } from '../check'
import { revert } from '../revert'
import { version } from './../../package.json'
import { generateXlsx } from './genXlsx'
import { init } from './init'
import { trans } from './trans'
import { updateLocales, updateLocalesFromXlsx } from './update'

const program = new Command()

program
  .name(CLI_CONFIG_NAME)
  .description('A CLI to help you transform your project '
    + 'to an internationalization project automatically.')
  .version(version)

program.addOption(new Option('-d, --debug').hideHelp())
program.addOption(new Option('-vb, --verbose').hideHelp())

program.on('option:verbose', () => {
  process.env.AUTO_I18N_VERBOSE = program.opts().verbose
})
program.on('option:debug', () => {
  process.env.AUTO_I18N_DEBUG = program.opts().debug
})

program.command('init')
  .description('Init yo-auto-i18n, generate an new default config file automatically.')
  .action(() => {
    init()
  })

program.command('trans')
  .description('Transform a single file or all files.')
  .argument('[transPath]', 'file or directory path')
  .option('--modify', 'whether to modify on the source file')
  .option('-t, --template <string>', 'template json file which inclues all language key')
  .action((transPath, options) => {
    const transCommandOption: TransCommandOption = {
      transPath: transPath ?? '',
      modifyMode: !!options.modify,
      templateFile: options.template ?? '',
    }
    trans(transCommandOption)
  })

program.command('update')
  .description('Update other locales Json files by base locales Json file')
  .option('-t, --templateXlsx <string>', 'update locales json files by the Xlsx template file,'
    + ' the frirt line of the first sheet should have all locales names and the `key`', '')
  .action(async (options) => {
    await updateLocales()
    options.templateXlsx && (await updateLocalesFromXlsx(options.templateXlsx))
  })

program.command('genlsx')
  .description('Generate an Xlsx file by locales json files,'
    + ' the frirt line of in the first sheet of the xlsx will have all locales-name and the `key`')
  .action(() => {
    generateXlsx()
  })

program.command('check')
  .description('Check weather all word has been transformed or not')
  .action(() => {
    checkAllTranslated()
  })

program.command('revert')
  .description('revert a transformed file or files in a directory')
  .argument('<revertPath>', 'file or directory path')
  .option('-t, --target <string>', 'the target language when revert')
  .action(async (revertPath, options) => {
    const revertCommandOption: RevertCommandOption = {
      revertPath: revertPath ?? '',
      target: options.target ?? '',
    }
    await revert(revertCommandOption)
  })

program.parse()
