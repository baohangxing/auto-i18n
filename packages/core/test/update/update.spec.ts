import path from 'path'
import fs from 'fs'
import { describe, expect, it } from 'vitest'
import { updateLocales } from '../../src/command/update'
import { addTestPlayground, getTestPlaygroundRootPath, removeTestPlayground } from '../testUtils/playground'

describe('#update', async () => {
  it('should update json file', async () => {
    addTestPlayground('update-1')
    const jsonPath = path.join(getTestPlaygroundRootPath('update-1'), 'locales', 'en.json')
    await updateLocales()
    const json = fs.readFileSync(jsonPath).toString()
    removeTestPlayground('update-1')
    expect(json).toMatchSnapshot()
  })
})
