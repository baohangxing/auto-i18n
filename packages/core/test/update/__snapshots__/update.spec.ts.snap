// Vitest Snapshot v1

exports[`#update > should update json file 1`] = `
"{
  \\"dazigao\\": \\"typescript\\",
  \\"title\\": \\"[EN]标题\\"
}
"
`;
